### PR TITLE
test:  add unit tests for analytics functions

### PR DESCRIPTION
**Pull request by ArkadiuszKrzymienski**
add unit tests for analytics function

### CODE REVIEW

File: tests/index.test.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    test(`${fn} - run all modules with params`, () => {
+      analytics.Analytics[fn](testObj);
+      analyticsModules.default.forEach((singleModule) => {
+        if (singleModule[fn]) {
+          expect(singleModule[fn]).toHaveBeenCalledWith(testObj);
+        }
+      });
+    });
+  });
+  test('Vue Analytics - add $analytics to vue prototype', () => {
```
Comment (pspaczek) [inline tests/index.test.js:83]: R: Maybe it should be renamed to something like `use as Plugin in Vue 2` WDYT?

File: tests/index.test.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      analytics.Analytics[fn](testStr, testObj, moduleNames);
+      analyticsModules.default.forEach((singleModule) => {
+        if (singleModule[fn]) {
+          expect(singleModule[fn]).toHaveBeenCalledWith(testStr, { global: 'asObject', ...testObj });
+        }
+      });
+    });
+  });
+  describe.each(['initialize', 'trackConversion'])('initialize and trackConversion function', (fn) => {
+    test(`${fn} - run all modules without params`, () => {
```
Comment (pspaczek) [inline tests/index.test.js:66]: k, this test looks fine you
1. trigger this function
2. and expect called for all modules
fine, for with params same 👍

File: tests/index.test.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      analytics.Analytics.setGlobalProperties({ global: 'asObject' });
+      analytics.Analytics[fn](testStr, testObj, moduleNames);
+      analyticsModules.default.forEach((singleModule) => {
+        if (singleModule[fn]) {
+          expect(singleModule[fn]).toHaveBeenCalledWith(testStr, { global: 'asObject', ...testObj });
+        }
+      });
+    });
+  });
+  describe.each(['initialize', 'trackConversion'])('initialize and trackConversion function', (fn) => {
```
[the remark this answers is not in the document]
Comment (pspaczek) [reply tests/index.test.js:65]: I know why, because is looks the same ;) but IMO they have a different responsibility ;)

File: tests/index.test.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    test(`${fn} - with setGlobalProperties as string, run all modules`, () => {
+      analytics.Analytics.setGlobalProperties('global', 'asString');
+      analytics.Analytics[fn](testStr, testObj, moduleNames);
+      analyticsModules.default.forEach((singleModule) => {
+        if (singleModule[fn]) {
+          expect(singleModule[fn]).toHaveBeenCalledWith(testStr, { global: 'asString', ...testObj });
+        }
+      });
+    });
+    test(`${fn} - with setGlobalProperties as object, run all modules`, () => {
```
Comment (pspaczek) [inline tests/index.test.js:55]: R: test `setGlobalProperties` in sandbox outside `trackView` and `trackEvent`. Test it for string, object. Just compare `globalProperties` before and after `setGlobalProperties`

File: tests/index.test.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    Object.keys(singleModule).forEach((key) => {
+      if (key === 'name') return;
+      jest.spyOn(singleModule, key);
+    });
+  });
+});
+
+describe('analytics function', () => {
+  describe.each(['trackEvent', 'trackView'])('trackView and trackEvent function', (fn) => {
+    test(`${fn} - run all modules`, () => {
```
Comment (pspaczek) [inline tests/index.test.js:26]: 👍

File: tests/index.test.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      analytics.Analytics[fn](testStr, testObj, moduleNames);
+      analyticsModules.default.forEach((singleModule) => {
+        if (singleModule[fn]) {
+          expect(singleModule[fn]).toHaveBeenCalledWith(testStr, testObj);
+        }
+      });
+    });
+    test(`${fn} - run only passed modules`, () => {
```
Comment (pspaczek) [inline tests/index.test.js:34]: 👍  but can see why mockups look better than existing modules. Now, this test is connected to modules.js... if you change something with names then it will be broken, but... why, it should be passes ;)

File: tests/index.test.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        if (singleModule[fn]) {
+          const isCalled = singleModule.name !== 'googleTagManager' && singleModule.name !== 'infermedicaAnalytics';
+          // eslint-disable-next-line no-unused-expressions
+          isCalled
+            ? expect(singleModule[fn]).toHaveBeenCalledWith(testStr, testObj)
+            : expect(singleModule[fn]).not.toHaveBeenCalled();
+        }
+      });
+    });
+    test(`${fn} - with setGlobalProperties as string, run all modules`, () => {
```
Comment (pspaczek) [inline tests/index.test.js:46]: 👍   but we need to test `setGlobalProperties` without trackEvent and `trackView` and we should name call this test `with global properties as string` ;) because here we test there result of trackView... not setGlobalProperties function

File: tests/index.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pspaczek) [reply tests/index.test.js:65]: maybe we should test  `await analytics.Analytics.initialize()` with some some setTimeOut in mock module to check what it be do when you want 
```
analytics.Analytics.initialize();
trackEvent(); // track event should be send to module with some delay 
```
because @AlanLes has problem with initialize in Intake and IMO it's useful test

**Review by pspaczek — CHANGES_REQUESTED** [review]
This PR is really impressive 🥇. But IMO we should change the approach to test analytics functions with mock instead existing modules. And we can add/ update some tests to cover edge cases. :thank_you: @ArkadiuszKrzymienski

File: tests/index.test.js [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+  return modules;
+};
+const setAnalyticModulesMock = async (moduleRules) => {
+  jest.resetModules();
+  jest.unmock('../src/modules');
+  jest.doMock('../src/modules', () => ({ __esModule: true, default: createModuleList(moduleRules) }));
+  analyticModules = await import('../src/modules');
+  analytics = await import('../src/index');
+  // eslint-disable-next-line prefer-destructuring
+  mockedAnalyticModules = analyticModules.default;
```
Comment (pspaczek) [inline tests/index.test.js:56]: R: and you can remove `// eslint-disable-next-line prefer-destructuring`
```js
const { default: mockedAnalyticModules } = analyticModules;

File: tests/index.test.js [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+      analytics.Analytics.trackEvent(testStr, testObj, 'notThisModuleName');
+      expect(mockedAnalyticModules[0].trackEvent).not.toHaveBeenCalledWith(testStr, testObj);
+    });
+    test('run trackEvent in module with global properties as string', async () => {
+      analytics.Analytics.setGlobalProperties('global', 'asString');
+      analytics.Analytics.trackEvent(testStr, testObj, moduleNames);
+      expect(mockedAnalyticModules[0].trackEvent).toHaveBeenCalledWith(testStr, { global: 'asString', ...testObj });
+    });
+    test('run trackEvent in module with global properties as object', async () => {
+      analytics.Analytics.setGlobalProperties({ global: 'asObject' });
```
Comment (pspaczek) [inline tests/index.test.js:78]: Q: typeof 'asObject' is string 🤔  or did I miss something?

File: tests/index.test.js [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    });
+    test('trackEvent is not throw error when module does not support trackEvent', async () => {
+      await setAnalyticModulesMock([trackViewModule]);
+      expect(() => analytics.Analytics.trackEvent(testStr, testObj, moduleNames)).not.toThrow(Error);
+    });
+    test('run trackEvent in all modules', async () => {
+      modulesSchema = [trackEventModule, trackEventModule, trackEventModule];
+      await setAnalyticModulesMock(modulesSchema);
+      analytics.Analytics.trackEvent(testStr, testObj, moduleNames);
+      expect(mockedAnalyticModules[0].trackEvent).toHaveBeenCalledWith(testStr, testObj);
```
Comment (pspaczek) [inline tests/index.test.js:95]: Q: maybe we can use `moduleSchema` array to iterable it with expect?
```
moduleSchema.forEach((module, index) => {
  expect(mockedAnalyticModules[index].trackEvent).toHaveBeenCalledWith(testStr, testObj);
})
```

File: tests/index.test.js [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    });
+    test('run trackEvent in only passed modules', async () => {
+      modulesSchema = [trackEventModule, trackEventModule, trackEventModule];
+      await setAnalyticModulesMock(modulesSchema);
+      analytics.Analytics.trackEvent(testStr, testObj, supportedModuleNames);
+      expect(mockedAnalyticModules[0].trackEvent).toHaveBeenCalledWith(testStr, testObj);
+      expect(mockedAnalyticModules[1].trackEvent).not.toHaveBeenCalledWith(testStr, testObj);
+      expect(mockedAnalyticModules[2].trackEvent).toHaveBeenCalledWith(testStr, testObj);
+    });
+
```
Comment (pspaczek) [inline tests/index.test.js:112]: R: empty line

File: tests/index.test.js [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    });
+    test('trackView is not throw error when module does not support trackView', async () => {
+      await setAnalyticModulesMock([trackEventModule]);
+      expect(() => analytics.Analytics.trackView(testStr, testObj, moduleNames)).not.toThrow(Error);
+    });
+    test('run trackView in all modules', async () => {
+      modulesSchema = [trackViewModule, trackViewModule, trackViewModule];
+      await setAnalyticModulesMock(modulesSchema);
+      analytics.Analytics.trackView(testStr, testObj, moduleNames);
+      expect(mockedAnalyticModules[0].trackView).toHaveBeenCalledWith(testStr, testObj);
```
Comment (pspaczek) [inline tests/index.test.js:173]: modulesSchema.forEach or other solution to avoid duplication 🙏

File: tests/index.test.js [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+  });
+  describe('setGlobalProperties function', () => {
+    beforeEach(async () => {
+      jest.resetModules();
+      jest.unmock('../src/modules');
+      analyticModules = await import('../src/modules');
+      analytics = await import('../src/index');
+    });
+    test('setGlobalProperties - as string', () => {
+      // eslint-disable-next-line
```
Comment (pspaczek) [inline tests/index.test.js:277]: R: when you use // eslint-disable... then please try to write what you disable

File: tests/index.test.js (hunk 9 above)
Comment (ArkadiuszKrzymienski) [reply tests/index.test.js:78]: Important is type of parameter passed to the setGlobalProperties function. In this case, it is object. I will rename the test to make it clear 😉

Comment (pspaczek) [reply tests/index.test.js:77]: 🙈  fine I miss brackets, :thank_you:

**Review by pspaczek — APPROVED** [review]
LGTM!
